x/interchainqueries/keeper/msg_server.go

```diff
@@ -0,0 +1,329 @@
+	if err != nil {
+		ctx.Logger().Error("SubmitQueryResult: failed to decode AccAddressFromBech32",
+			"error", err, "query", query, "message", msg)
+		return nil, sdkerrors.Wrapf(err, "failed to decode owner contract address (%s)", query.Owner)
+	}
+
+	if msg.Result.KvResults != nil {
+		if !types.InterchainQueryType(query.QueryType).IsKV() {
+			return nil, sdkerrors.Wrapf(types.ErrInvalidType, "invalid query result for query type: %s", query.QueryType)
+		}
+		if err := k.checkLastRemoteHeight(ctx, *query, ibcclienttypes.NewHeight(msg.Result.Revision, msg.Result.Height)); err != nil {
+			return nil, sdkerrors.Wrap(types.ErrInvalidHeight, err.Error())
+		}
+		if len(msg.Result.KvResults) != len(query.Keys) {
+			return nil, sdkerrors.Wrapf(types.ErrInvalidSubmittedResult, "KV keys length from result is not equal to registered query keys length: %v != %v", len(msg.Result.KvResults), len(query.Keys))
+		}
+
+		clientState, err := k.GetClientState(ctx, msg.ClientId)
+		if err != nil {
+			return nil, err
+		}
+
+		for index, result := range msg.Result.KvResults {
+			proof, err := ibccommitmenttypes.ConvertProofs(result.Proof)
+			if err != nil {
+				ctx.Logger().Debug("SubmitQueryResult: failed to ConvertProofs",
+					"error", err, "query", query, "message", msg)
+				return nil, sdkerrors.Wrapf(types.ErrInvalidType, "failed to convert crypto.ProofOps to MerkleProof: %v", err)
+			}
+
+			if !bytes.Equal(result.Key, query.Keys[index].Key) {
+				return nil, sdkerrors.Wrapf(types.ErrInvalidSubmittedResult, "KV key from result is not equal to registered query key: %v != %v", result.Key, query.Keys[index].Key)
+			}
+
+			if result.StoragePrefix != query.Keys[index].Path {
+				return nil, sdkerrors.Wrapf(types.ErrInvalidSubmittedResult, "KV path from result is not equal to registered query storage prefix: %v != %v", result.StoragePrefix, query.Keys[index].Path)
+			}
+
+			path := ibccommitmenttypes.NewMerklePath(result.StoragePrefix, url.PathEscape(string(result.Key)))
+			key, err := path.GetKey(uint64(len(path.KeyPath) - 1))
+			if err != nil {
+				return nil, sdkerrors.Wrapf(ibccommitmenttypes.ErrInvalidProof, "could not retrieve key bytes for key: %s", path.KeyPath[len(path.KeyPath)-1])
+			}
+
+			subroot, err := proof.Proofs[0].Calculate()
+
+			// identify what kind proofs (non-existence proof always has *ics23.CommitmentProof_Nonexist as the first item) we got
+			// and call corresponding method to verify it
+			switch proof.GetProofs()[0].Proof.(type) {
+			// we can get non-existence proof if someone queried some key which is not exists in the storage on remote chain
+			case *ics23.CommitmentProof_Nonexist:
+
+				if err != nil {
+					return nil, sdkerrors.Wrapf(ibccommitmenttypes.ErrInvalidProof, "could not calculate root for proof index 0, merkle tree is likely empty. %v", err)
+				}
+				if ok := ics23.VerifyNonMembership(clientState.ProofSpecs[0], subroot, proof.Proofs[0], key); !ok {
```

Comment:
Changed this from `proof.VerifyNonMembership` (which was in the merkle proof package) to ics23 `VerifuNonMembership`, need double confirmation this is ok